DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,688 @@
+By making `bool` an unsafe type, it can be both fast and `@safe` since all tricks to make invalid booleans are now disabled in `@safe` code.
+The cost is that existing code might break.
+Note that with the proposed rules a `bool` variable or field is only `@system` when the compiler cannot possibly prove that the initial value is 0 or 1, which should be a rare occurrence.
+If there is code breakage, it will likely come from overlap or void initialization of booleans.
+```D
+immutable ubyte ub = 3;
+bool getValidBool()   pure @system {return true;}
+bool getInvalidBool() pure @system {return *(cast(bool*) ub);}
+
+bool b0 = getValidBool();   // despite unsafe type with @system initialization expression, inferred as @safe
+bool b1 = getInvalidBool(); // inferred as system
+
+void main() @safe {
+    bool b = void; // no longer compiles
+}
+```
+
+### Grammar changes
+
+There are no proposed grammar changes, since placing `@system` annotations is already allowed on the places where it's needed for this DIP.
+
+## Examples
+
+This section contains five examples that demonstrate use cases for the proposed changes.
+Each example *wrongly uses `@trusted`* under *current semantics* followed by a `main` function demonstrating why.
+After this DIP is implemented, these will all become correct usages of `@trusted` and the `main` functions will fail to compile.
+
+### A tagged union
+This example shows how to safely overlap an unsafe type with another type.
+```D
+struct Var {
+    private @system union {
+        string asString = "var";
+        double asDouble;
+    }
+    enum Type : byte {
+        Str,
+        Num,
+    }
+    private @system Type type;
+
+    this(double x) @trusted {
+        this.type = Type.Num;
+        this.asDouble = x;
+    }
+
+    this(string s) @trusted {
+        this.type = Type.Str;
+        this.asString = s;
+    }
+
+    double getDouble() const @trusted {
+        return type == Type.Str ? asDouble : double.nan;
+    }
+
+    string getString() const @trusted {
+        return type == Type.Str ? asString : null;
+    }
+}
+
+import std;
+
+void main() @safe {
+    Var v = 1.0;
+    writeln(v.asString); // memory corruption: the string's length (and pointer
+    // on 32-bit targets) have been meddled with
+}
+```
+
+### Using uninitialized memory
+This simple stack data structure shows how to safely use uninitialized memory.
+In this case the stack data structure is using uninitialized stack memory, but the same idea applies to using malloc'ed memory.
+```D
+private struct Stack(T, int size = 32) {
+    private @system T[size] stack = void;
+    private @system int _top = 0;
+
+    T pop() @trusted {if (empty) assert(0); return stack[--_top];}
+    void push(T x) @trusted {if (full) assert(0); stack[_top++] = x;}
```

Comment:
This is somewhat besides the point of the DIP, but these methods can't be `@trusted` for other reasons. `pop` and `push` will call an unsafe copy constructor or `opAssign`:

```d
struct S
{
    import core.stdc.stdio: printf;
    void opAssign(ref S src) @system { printf("unsafe opAssign\n"); }
    this(ref S src) @system { printf("unsafe copy ctor\n"); }
}
void main() @safe
{
    auto stack = getStack!S();
    stack.push(S()); /* calls unsafe opAssign */
    auto s = stack.pop(); /* calls unsafe copy ctor */
}
```

Also, `T` might contain a pointer and it might have a destructor that writes through that pointer. That can't be allowed to run on uninitialized memory.

```d
struct S
{
    int* p;
    ~this() @safe { *p = 42; }
}
void f() @safe { auto stack = getStack!S(); }
    /* writes 42 to undefined memory location */
```

Automatically adding an `@system` destructor might help here.

Reply:
While indeed besides the point of the DIP, I do find it important that the examples are correct, so it's good you point these things out. I simply added the restriction `isPOD` to keep the example simple and limited the scope of `@trusted` for `push`.

> Automatically adding an @system destructor might help here.

I don't know what you mean with that.

Reply:
> I don't know what you mean with that.

Never mind. I had missed a couple things there.

The idea was to infer `Stack`'s implicit destructor as `@system` so that `Stack` can't be used in `@safe` code. Then unsafe destruction would be prevented.

I thought that you could then write an explicit `@trusted` destructor that avoids destruction of uninitialized data. But even with an explicit destructor, it seems that the destructors of fields are still called automatically. So unsafe destruction can't be prevented that way.

And anyway, it probably makes more sense to say that `getStack` violates safety if the returned object can't be destroyed safely.

With the `isPOD` restriction, all this doesn't apply anymore, of course.